enhancements/microshift/microshift-router-configuration.md

```diff
@@ -0,0 +1,621 @@
+deployed. As part of the start procedure, MicroShift will configure the router
+with fixed parameters, such as being always enabled, and tied to ports 80 and
+443. None of these are configurable for the user, and they also tie to how the
+router is exposed in the cluster.
+
+In order to allow more flexibility, the following configuration parameters shall
+be added:
+- Enable/Disable. Some use cases may be egress only, meaning they do not need
+  an ingress controller. In this case the router pod should not be scheduled
+  to save resources. There should be no rules associated to firewalld or
+  iptables either for an improved security posture.
+- Listening ports. Allow configuring ports other than 80 and 443 improves
+  flexibility.
+- Which IPs the router listens on. Some use cases may require the router to be
+  reachable only from certain networks.
+
+### User Stories
+As a MicroShift user, I want to enable/disable the default router.
+
+As a MicroShift user, I want to disable the default router.
+
+As a MicroShift user, I want to be able to configure in which ports the default
+router is listening.
+
+As a MicroShift user, I want to be able to choose the IPs where the router is
+listening.
+
+As a MicroShift user, I want to be able to allow traffic to the router from
+specific IP addresses.
+
+As a MicroShift user, I want to be able to deny traffic to the router from
+specific IP addresses.
+
+### Goals
+* Allow users to enable the router.
+* Allow users to disable the router.
+* Allow users to reach the router in the configured ports.
+* Allow users to configure in which IPs the router listens.
+* Allow users to allow traffic from specific IP addresses.
+* Allow users to deny traffic from specific IP addresses.
+
+### Non-Goals
+* N/A
+
+## Proposal
+Each configuration option will get its own subsection with specifics and
+details.
+As a common baseline, a new top level section will be added to the
+configuration:
+```yaml
+apiServer:
+   ...
+# new from here.
+router:
+  enable: <true|false> # Defaults to true.
+  policy:
+    ports:
+      http:
+        enabled: <true|false> # Defaults to false.
+        port: <int> # Defaults to 80.
+      https:
+        enabled: <true|false> # Defaults to true.
+        port: <int> # Defaults to 443.
+    expose:
+      - <IP address/NIC> # Defaults to the NICs of the node
+    allow:
+      - <IP address> # Defaults to empty.
+```
+
+### Enable/disable the router
+The following configuration is proposed:
+```yaml
+router:
+  enable: <true|false> # Defaults to true.
+```
+
+With this option MicroShift will decide whether it should create the router
+upon starting. This includes not just the pod, but all the associated resources
+that come along with it: namespace, services, configmaps, etc.
+
+When setting the option to `false`, the next MicroShift restart should delete
+all the router resources, if any were created before and relate to the router
+deployment, but not the RBAC resources.
```

Comment:
We should still allow route API resources, which involves RBAC. This proposal is about removing the default router, but not the whole API.

There might be use cases where you would need to run the router under specific parameters/conditions, which may be different from the default router. This would require having the API, controllers, RBAC, etc.

Reply:
For now let's assume it's all or nothing. We can list as a risk that there would be rework in the future if someone wanted to disable the default router and configure a custom router. We can deal with that if the feature request comes up.

Reply:
That sounds reasonable and is also easier to do. It is included now.